### PR TITLE
Fix incorrect predicate push down with grouping sets

### LINKED ISSUE #45947: GROUPNG SETS with predicate return empty result.
How to reproduce:
```sql
drop table if exists test.grouping;
create  table test.grouping
(
    `day_` Date,
    `type_1` String,
)
ENGINE = MergeTree
ORDER BY day_ ;

INSERT INTO test.grouping SELECT
    toDate('2023-01-05') AS day_,
    randomString(10)
FROM numbers(1);
```

query:
```sql
SELECT                                                                                                                                                                                                              
    day_,                                                                                                                                                                                                           
    if(type_1 = '', 'all', type_1) AS type_1                                                                                                                                                                        
FROM                                                                                                                                                                                                                
(                                                                                                                                                                                                                   
    SELECT                                                                                                                                                                                                          
        day_,                                                                                                                                                                                                       
        type_1                                                                                                                                                                                                      
    FROM test.grouping                                                                                                                                                                                              
    WHERE day_ = '2023-01-05'                                                                                                                                                                                       
    GROUP BY                                                                                                                                                                                                        
        GROUPING SETS (                                                                                                                                                                                             
            (day_, type_1),                                                                                                                                                                                         
            (day_))                                                                                                                                                                                                 
) AS t                                                                                                                                                                                                              
                                                                                                                                                                                                                    
Query id: 0639e9c0-4b75-414d-93b6-f9f9079740d3                                                                                                                                                                      
                                                                                                                                                                                                                    
┌───────day_─┬─type_1─┐                                                                                                                                                                                             
│ 2023-01-05 │ /4{                                                                                                                                                                                                  
fQ  │                                                                                                                                                                                                               
└────────────┴────────┘                                                                                                                                                                                             
┌───────day_─┬─type_1─┐                                                                                                                                                                                             
│ 2023-01-05 │ all    │                                                                                                                                                                                             
└────────────┴────────┘                                                                                                                                                                                             
                                                                                                                                                                                                                    
2 rows in set. Elapsed: 0.003 sec.    
```

```sql
SELECT *
FROM
(
    SELECT
        day_,
        if(type_1 = '', 'all', type_1) AS type_1
    FROM
    (
        SELECT
            day_,
            type_1
        FROM test.grouping
        WHERE day_ = '2023-01-05'
        GROUP BY
            GROUPING SETS (
                (day_, type_1),
                (day_))
    ) AS t
)
WHERE type_1 = 'all'

Query id: 00b6ffcc-76c0-48bc-a45a-57b5c593439d

Ok.

0 rows in set. Elapsed: 0.006 sec. 
```
This query should return one row, but nothing return.

### PR DESCRIPTION
**Pull request by ucasfl**
### Changelog category (leave one):

- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix incorrect predicate push down with grouping sets. Closes https://github.com/ClickHouse/ClickHouse/issues/45947

### CODE REVIEW

File: src/Interpreters/PredicateExpressionsOptimizer.cpp [hunk 1]
```diff
@@ -35,7 +35,9 @@ bool PredicateExpressionsOptimizer::optimize(ASTSelectQuery & select_query)
     if (!enable_optimize_predicate_expression)
         return false;
 
-    if (select_query.having() && (!select_query.group_by_with_cube && !select_query.group_by_with_rollup && !select_query.group_by_with_totals))
+    if (select_query.having()
+        && (!select_query.group_by_with_cube && !select_query.group_by_with_rollup && !select_query.group_by_with_totals
+            && !select_query.group_by_with_grouping_sets))
```
Comment (nickitat) [inline src/Interpreters/PredicateExpressionsOptimizer.cpp:40]:
```suggestion
    const bool has_incompatible_constructs = select_query.group_by_with_cube || select_query.group_by_with_rollup || select_query.group_by_with_totals || select_query.group_by_with_grouping_sets;
    if (select_query.having() && !has_incompatible_constructs)
```

File: src/Processors/QueryPlan/Optimizations/filterPushDown.cpp [hunk 2]
```diff
@@ -176,6 +176,9 @@ size_t tryPushDownFilter(QueryPlan::Node * parent_node, QueryPlan::Nodes & nodes
 
     if (auto * aggregating = typeid_cast<AggregatingStep *>(child.get()))
     {
+        if (aggregating->isGroupingSets())
+            return 0;
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/filterPushDown.cpp:180]: In this case we can apply optimisation if all identifiers in the predicate are use in each grouping set

**Review by novikd — APPROVED** [review]

File: src/Processors/QueryPlan/Optimizations/filterPushDown.cpp [hunk 3]
```diff
@@ -53,6 +53,53 @@ static void checkChildrenSize(QueryPlan::Node * node, size_t child_num)
...
+            if (std::find(grouping_set.used_keys.begin(), grouping_set.used_keys.end(), identifier) == grouping_set.used_keys.end())
+                return false;
+        }
+    }
+    return true;
+}
+
+static Names findIdentifiersOfNode(const ActionsDAG::Node * node)
+{
+    Names res;
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/filterPushDown.cpp:71]: Better to use `NameSet`.

Comment (ucasfl) [reply src/Processors/QueryPlan/Optimizations/filterPushDown.cpp]: done.

**Comment by nickitat** [comment]
test failures seem irrelevant

**Review by nickitat — APPROVED** [review]

**Pull request merged**